zaza/openstack/utilities/parallel_series_upgrade.py

```diff
@@ -0,0 +1,639 @@
+                                   on all machine in the application
+    :param follower_first: Should the follower(s) be upgraded first
+    :type follower_first: bool
+    :type post_application_upgrade_functions: List[str]
+    :param files: Workaround files to scp to unit under upgrade
+    :type files: list
+    :param workaround_script: Workaround script to run during series upgrade
+    :type workaround_script: str
+    :returns: None
+    :rtype: None
+    """
+    if completed_machines is None:
+        completed_machines = []
+    if follower_first:
+        logging.error("leader_first is ignored for parallel upgrade")
+    logging.info(
+        "About to upgrade the units of {} in parallel (follower first: {})"
+        .format(application, follower_first))
+
+    status = (await model.async_get_status()).applications[application]
+    logging.info(
+        "Configuring leader / non leaders for {}".format(application))
+    leader, non_leaders = get_leader_and_non_leaders(status)
+    for leader_name, leader_unit in leader.items():
+        leader_machine = leader_unit["machine"]
+        leader = leader_name
+    machines = [
+        unit["machine"] for name, unit
+        in non_leaders.items()
+        if unit['machine'] not in completed_machines]
+
+    await maybe_pause_things(
+        status,
+        non_leaders,
+        pause_non_leader_subordinate,
+        pause_non_leader_primary)
+    await series_upgrade_utils.async_set_series(
+        application, to_series=to_series)
+    app_idle = [
+        wait_for_unit_idle(unit) for unit in status["units"]
+    ]
+    await asyncio.gather(*app_idle)
+    prepare_group = [
+        prepare_series_upgrade(machine, to_series=to_series)
+        for machine in machines]
+    await asyncio.gather(*prepare_group)
+    await prepare_series_upgrade(leader_machine, to_series=to_series)
+    if leader_machine not in completed_machines:
+        machines.append(leader_machine)
+    upgrade_group = [
+        series_upgrade_machine(
+            machine,
+            origin=origin,
+            application=application,
+            files=files, workaround_script=workaround_script,
+            post_upgrade_functions=post_upgrade_functions)
+        for machine in machines
+    ]
+    await asyncio.gather(*upgrade_group)
+    completed_machines.extend(machines)
+    await run_post_application_upgrade_functions(
+        post_application_upgrade_functions)
+
+
+async def serial_series_upgrade(
+    application,
+    from_series='xenial',
+    to_series='bionic',
+    origin='openstack-origin',
+    pause_non_leader_primary=True,
+    pause_non_leader_subordinate=True,
+    pre_upgrade_functions=None,
+    post_upgrade_functions=None,
+    post_application_upgrade_functions=None,
+    completed_machines=None,
+    follower_first=False,
+    files=None,
+    workaround_script=None
+):
+    """Perform series upgrade on an application in parallel.
```

Comment:
s/parallel/series/

Reply:
This is performing a series upgrade, in parallel